app/models/cluster_group.rb

```diff
@@ -312,7 +313,19 @@ def generate_subsample_arrays(sample_size, annotation_name, annotation_type, ann
 
   # determine which subsampling levels are required for this cluster
   def subsample_thresholds_required
-    ClusterGroup::SUBSAMPLE_THRESHOLDS.select {|sample| sample < self.points}
+    SUBSAMPLE_THRESHOLDS.select {|sample| sample < self.points}
+  end
+
+  # control gate for invoking subsampling
+  def can_subsample?
+    if self.points < SUBSAMPLE_THRESHOLDS.min || self.subsampled
+      false
+    else
+      # check if there are any data arrays belonging to this cluster that have a subsample threshold & annotation
+      !DataArray.where(study_id: self.study_id, study_file_id: self.study_file_id, linear_data_type: 'ClusterGroup',
```

Comment:
seeing this pattern repeat itself made me debate with myself if it would be worth method extracting something like this:
`
!matching_data_arrays(self).any?
`
`
def self.matching_data_arrays(cluster_group)
  DataArray.where(study_id: self.study_id, study_file_id: self.study_file_id, linear_data_type: 'ClusterGroup',
    linear_data_id: self.id, :subsample_threshold.nin => [nil],
                       :subsample_annotation.nin => [nil])
end
`
..with delete_queue_job taking advantage as well.

Reply:
Good call - I'll go ahead and use that.